README.md
# flutter_health

This library combines both GoogleFit and AppleHealthKit. It support most of the values provided.

Works on from **iOS 11.0**. Some data types are supported from **iOS 12.2**.

Supports **Android X**


[Dart package](https://pub.dev/packages/flutter_health)

[Gitlab link](https://gitlab.com/petleo-and-iatros-opensource/flutter_health)


# AppleHealthKit
First Let's go through apple health kit:

## Setup

From Runner, select the **Capabilities tab** 

* Enable **HealthKit** 
* Check the **Clinical Health Records** box.

Then in the info.plist, Add string values for the following:

* **NSHealthUpdateUsageDescription**
* **NSHealthShareUsageDescription**

## How to use

#### Check if Apple Health is available on the device

```$xslt
await FlutterHealth.checkIfHealthDataAvailable()
```

#### Request authorization for the data types supported by the plugin

```$xslt
await FlutterHealth.requestAuthorization()
``` 
###### For now, you can request all of the options provided by the library


#### Get the samples of data types authorized by the user within the given time range

```$xslt
await FlutterHealth.getBodyFat(startDate, endDate)
``` 

```$xslt
await FlutterHealth.getHeartRate(startDate, endDate)
``` 

```$xslt
await FlutterHealth.getBloodPressureSys(startDate, endDate)
``` 
...



#### Data will be returning as a List of an object called HKHealthData. This object has the following attributes:

```$xslt
 double value;
 String unit;
 int dateFrom;
 int dateTo;
``` 

### You can get those values with this library:

**Tested:** 

* bodyFatPercentage
* height
* bodyMassIndex
* waistCircumference
* stepCount
* basalEnergyBurned
* activeEnergyBurned
* heartRate
* restingHeartRate
* walkingHeartRateAverage
* bodyTemperature
* bloodPressureSystolic
* bloodPressureDiastolic
* oxygenSaturation
* bloodGlucose
* electrodermalActivity 

**Could not be tested without a watch:**
  
* highHeartRateEvent
* lowHeartRateEvent
* irregularHeartRhythmEvent

## Full Example

```$xslt
 await FlutterHealth.checkIfHealthDataAvailable();
 bool _isAuthorized = true;
 DateTime startDate = DateTime.utc(2018);
 DateTime endDate = DateTime.now();
 var _dataList = List<HKHealthData>();
 
 _isAuthorized = await FlutterHealth.requestAuthorization();
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBodyFat(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHeight(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBodyMass(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getWaistCircumference(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getStepCount(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBasalEnergyBurned(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getActiveEnergyBurned(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHeartRate(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getRestingHeartRate(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getWalkingHeartRate(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBodyTemperature(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodPressureSys(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodPressureDia(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodOxygen(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodGlucose(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getElectrodermalActivity(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKHighHeart(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKLowHeart(startDate, endDate));
 if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKIrregular(startDate, endDate));

``` 




#GoogleFit


## Setup

For GoogleFit, the initial setup is a bit longer, and can get frustrating.
Just follow [this setup](https://developers.google.com/fit/android/get-started). 

## How to use

#### Request authorization for the data types supported by the plugin

```$xslt
await FlutterHealth.requestAuthorization()
``` 
###### For now, you can request all of the options provided by the library


#### Get the samples of data types authorized by the user within the given time range


For all data
```$xslt
await FlutterHealth.getGFAllData(startDate, endDate)
``` 

For specific type
```$xslt
await FlutterHealth.getGFBodyFat(startDate, endDate)
``` 

```$xslt
await FlutterHealth.getGFHeartRate(startDate, endDate)
``` 

```$xslt
await FlutterHealth.getGFBloodPressureSys(startDate, endDate)
``` 
...


#### Data will be returning as a List of an object called GFHealthData. This object has the following attributes:

```$xslt
 String value;
 String value2;
 String unit;
 int dateFrom;
 int dateTo;
``` 

### You can get those GoogleFit values with this library:

* bodyFatPercentage
* height
* stepCount
* calories
* heartRate
* bodyTemperature
* bloodPressure
* oxygenSaturation
* bloodGlucose

CHANGELOG.md
## 1.2.3
Adding Weight

## 1.2.1
Updating google fit and auth in gradle

## 1.2.0
Provide compatibility with Gradle 5.x

## 1.1.9

Apple health kit for flutter (Flutter Health) initial release

You can use it to get any of the following

Tested: 

* bodyFatPercentage
* height
* bodyMassIndex
* waistCircumference
* stepCount
* basalEnergyBurned
* activeEnergyBurned
* heartRate
* restingHeartRate
* walkingHeartRateAverage
* bodyTemperature
* bloodPressureSystolic
* bloodPressureDiastolic
* oxygenSaturation
* bloodGlucose
* electrodermalActivity 

Could not be tested:
  
* highHeartRateEvent
* lowHeartRateEvent
* irregularHeartRhythmEvent


test/flutter_health_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_health/flutter_health.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_health');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterHealth.platformVersion, '42');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_health/flutter_health.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var _healthKitOutput;
  var _dataList = List<HKHealthData>();
  var str = "";
  bool _isAuthorized = true;

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    /*Future.delayed(Duration(seconds: 2), () async {
      _healthKitOutput = await FlutterHealth.checkIfHealthDataAvailable();
      setState(() {});
    });*/

    DateTime startDate = DateTime.utc(2020, 03, 21);
    DateTime endDate = DateTime.now();
    Future.delayed(Duration(seconds: 2), () async {
      _isAuthorized = await FlutterHealth.requestAuthorization();
//      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKStepCountSummary(startDate, endDate));
//      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKAllDataWithCombinedBP(startDate, endDate));
      setState(() {});
//      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKStepCount(startDate, endDate));
//      setState(() {});
//      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHKHeartRate(startDate, endDate));
//      setState(() {});
      /* if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHeight(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBodyMass(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getWaistCircumference(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getStepCount(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBasalEnergyBurned(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getActiveEnergyBurned(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getHeartRate(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getRestingHeartRate(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getWalkingHeartRate(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBodyTemperature(startDate, endDate));
      setState(() {});
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodPressureSys(startDate, endDate));
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodPressureDia(startDate, endDate));
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodOxygen(startDate, endDate));
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getBloodGlucose(startDate, endDate));
      if (_isAuthorized) _dataList.addAll(await FlutterHealth.getElectrodermalActivity(startDate, endDate));
      setState(() {});*/
    });

    /*Future.delayed(Duration(seconds: 2), () async {
      _healthKitOutput = await FlutterHealth.getBloodType(context);
      setState(() {});
    });*/

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
          actions: <Widget>[
            IconButton(
              icon: Icon(Icons.sync),
              onPressed: () {
                initPlatformState();
              },
            )
          ],
        ),
        body: _dataList.isEmpty
            ? Text('Running on: $_healthKitOutput\n')
            : ListView.builder(
            itemCount: _dataList.length,
            itemBuilder: (_, index) =>
                ListTile(
                  title: Text(_dataList[index].value.toString() + " " + _dataList[index].value2.toString()),
                  trailing: Text(_dataList[index].unit),
                  subtitle: Text(DateTime.fromMillisecondsSinceEpoch(_dataList[index].dateFrom).toIso8601String()),
                  /*title: Text(str),
                  trailing: Text(str),
                  subtitle: Text(str),*/
                )),
      ),
    );
  }
}


android/src/main/kotlin/net/petleo/flutter_health/FlutterHealthPlugin.kt
package net.petleo.flutter_health

import android.app.Activity
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.fitness.Fitness
import com.google.android.gms.fitness.FitnessOptions
import com.google.android.gms.fitness.data.DataType
import com.google.android.gms.fitness.data.Field
import com.google.android.gms.fitness.data.HealthDataTypes
import com.google.android.gms.fitness.data.HealthFields
import com.google.android.gms.fitness.request.DataReadRequest
import com.google.android.gms.fitness.result.DataReadResponse
import com.google.android.gms.tasks.Tasks
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.ActivityResultListener
import io.flutter.plugin.common.PluginRegistry.Registrar
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread

const val GOOGLE_FIT_PERMISSIONS_REQUEST_CODE = 1111

class FlutterHealthPlugin(val activity: Activity, val channel: MethodChannel) : MethodCallHandler, ActivityResultListener, Result {

    private var result: Result? = null
    private var handler: Handler? = null

    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "flutter_health")
            val plugin = FlutterHealthPlugin(registrar.activity(), channel)

            registrar.addActivityResultListener(plugin)
            channel.setMethodCallHandler(plugin)
        }
    }


    fun MainThreadResult(result: Result) {
        this.result = result
        handler = Handler(Looper.getMainLooper())
    }

    override fun success(p0: Any?) {
        handler?.post(
                Runnable { result?.success(p0) })
    }

    override fun notImplemented() {
        handler?.post(
                Runnable { result?.notImplemented() })
    }

    override fun error(
            errorCode: String, errorMessage: String?, errorDetails: Any?) {
        handler?.post(
                Runnable { result?.error(errorCode, errorMessage, errorDetails) })
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Boolean {
        Log.d("FLUTTER_HEALTH", "GOOGLE FIT ON ACTIVITY RESULT $resultCode")
        if (requestCode == GOOGLE_FIT_PERMISSIONS_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                Log.d("FLUTTER_HEALTH", "Access Granted!")
                mResult?.success(true)
            } else {
                Log.d("FLUTTER_HEALTH", "Access Denied!")
            }
        }
        return false
    }

    var mResult: Result? = null

    val fitnessOptions = FitnessOptions.builder()
            .addDataType(DataType.TYPE_BODY_FAT_PERCENTAGE, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_HEIGHT, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_WEIGHT, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_STEP_COUNT_DELTA, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_CALORIES_EXPENDED, FitnessOptions.ACCESS_READ)
            .addDataType(DataType.TYPE_HEART_RATE_BPM, FitnessOptions.ACCESS_READ)
            .addDataType(HealthDataTypes.TYPE_BODY_TEMPERATURE, FitnessOptions.ACCESS_READ)
            .addDataType(HealthDataTypes.TYPE_BLOOD_PRESSURE, FitnessOptions.ACCESS_READ)
            .addDataType(HealthDataTypes.TYPE_OXYGEN_SATURATION, FitnessOptions.ACCESS_READ)
            .addDataType(HealthDataTypes.TYPE_BLOOD_GLUCOSE, FitnessOptions.ACCESS_READ)
            .build()


    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "requestAuthorization") {
            mResult = result

            if (!GoogleSignIn.hasPermissions(GoogleSignIn.getLastSignedInAccount(activity), fitnessOptions)) {
                Log.d("authResult 111", activity.localClassName)
                GoogleSignIn.requestPermissions(
                        activity, // your activity
                        GOOGLE_FIT_PERMISSIONS_REQUEST_CODE,
                        GoogleSignIn.getLastSignedInAccount(activity),
                        fitnessOptions)
            } else {
                mResult?.success(true)
                Log.d("FLUTTER_HEALTH", "Access already granted before!")
            }
        } else if (call.method == "getGFHealthData") {
            val type = call.argument<Int>("index")
            val startTime = call.argument<Long>("startDate")
            val endTime = call.argument<Long>("endDate")
            val fields = listOf(Field.FIELD_PERCENTAGE, Field.FIELD_HEIGHT, Field.FIELD_STEPS, Field.FIELD_CALORIES, Field.FIELD_BPM, HealthFields.FIELD_BODY_TEMPERATURE, HealthFields.FIELD_BLOOD_PRESSURE_SYSTOLIC, HealthFields.FIELD_OXYGEN_SATURATION, HealthFields.FIELD_BLOOD_GLUCOSE_LEVEL)
            val dataType = when (type) {
                0 -> DataType.TYPE_BODY_FAT_PERCENTAGE
                1 -> DataType.TYPE_HEIGHT
                2 -> DataType.TYPE_STEP_COUNT_DELTA
                3 -> DataType.TYPE_CALORIES_EXPENDED
                4 -> DataType.TYPE_HEART_RATE_BPM
                5 -> HealthDataTypes.TYPE_BODY_TEMPERATURE
                6 -> HealthDataTypes.TYPE_BLOOD_PRESSURE
                7 -> HealthDataTypes.TYPE_OXYGEN_SATURATION
                8 -> HealthDataTypes.TYPE_BLOOD_GLUCOSE
                9 -> DataType.TYPE_WEIGHT
                else -> DataType.TYPE_STEP_COUNT_DELTA
            }
            thread {
                val gsa = GoogleSignIn.getAccountForExtension(activity.applicationContext, fitnessOptions)

                val response = Fitness.getHistoryClient(activity.applicationContext, gsa)
                        .readData(DataReadRequest.Builder()
                                .read(dataType)
                                .setTimeRange(startTime ?: 0, endTime
                                        ?: 0, TimeUnit.MILLISECONDS)
                                .build())

                val readDataResult = Tasks.await<DataReadResponse>(response)
                val dataSet = readDataResult.getDataSet(dataType)

                val map = dataSet.dataPoints.map {
                    val map = HashMap<String, Any>()
                    map["value"] = try {
                        it.getValue(fields[type ?: 0]).asFloat()
                    } catch (e1: Exception) {
                        try {
                            it.getValue(fields[type ?: 0]).asInt()
                        } catch (e2: Exception) {
                            try {
                                it.getValue(fields[type ?: 0]).asString()
                            } catch (e3: Exception) {
                                Log.e("FLUTTER_HEALTH::ERROR", e3.toString())
                            }
                        }
                    }
                    if (dataType == HealthDataTypes.TYPE_BLOOD_PRESSURE)
                        map["value2"] = try {
                            it.getValue(HealthFields.FIELD_BLOOD_PRESSURE_DIASTOLIC).asFloat()
                        } catch (e1: Exception) {
                            try {
                                it.getValue(fields[type ?: 0]).asInt()
                            } catch (e2: Exception) {
                                try {
                                    it.getValue(fields[type ?: 0]).asString()
                                } catch (e3: Exception) {
                                    Log.e("FLUTTER_HEALTH::ERROR", e3.toString())
                                }
                            }
                        }
                    map["date_from"] = it.getStartTime(TimeUnit.MILLISECONDS)
                    map["date_to"] = it.getEndTime(TimeUnit.MILLISECONDS)
                    map["unit"] = ""
                    map["data_type_index"] = type ?: -1
                    return@map map
                }
                activity.runOnUiThread { result.success(map) }
            }

        } else {
            result.notImplemented()
        }
    }
}

ios/Classes/SwiftFlutterHealthPlugin.swift
import Flutter
import UIKit
import HealthKit

public class SwiftFlutterHealthPlugin: NSObject, FlutterPlugin {
    
    let healthStore = HKHealthStore()

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_health", binaryMessenger: registrar.messenger())
    let instance = SwiftFlutterHealthPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if(call.method.elementsEqual("checkIfHealthDataAvailable")){
        result(HKHealthStore.isHealthDataAvailable())
    } else if(call.method.elementsEqual("requestAuthorization")){
        var heartRateEventTypes = Set<HKSampleType>()
        if #available(iOS 12.2, *){
            heartRateEventTypes =  Set([
                HKSampleType.categoryType(forIdentifier: .highHeartRateEvent)!,
                HKSampleType.categoryType(forIdentifier: .lowHeartRateEvent)!,
                HKSampleType.categoryType(forIdentifier: .irregularHeartRhythmEvent)!,
               ])
//            healthStore.requestAuthorization(toShare: nil, read: heartRateEventTypes) { (success, error) in
//                if !success {
//                    result(false)// Handle the error here.
//                } else{
//                    result(true)
//                }
//            }
        }
        if #available(iOS 11.0, *) {
            let allTypes = heartRateEventTypes.union(Set([
                HKObjectType.workoutType(),
                HKSampleType.quantityType(forIdentifier: .bodyFatPercentage)!,
                HKSampleType.quantityType(forIdentifier: .height)!,
                HKSampleType.quantityType(forIdentifier: .bodyMassIndex)!,
                HKSampleType.quantityType(forIdentifier: .waistCircumference)!,
                HKSampleType.quantityType(forIdentifier: .stepCount)!,
                HKSampleType.quantityType(forIdentifier: .basalEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .activeEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .heartRate)!,
                HKSampleType.quantityType(forIdentifier: .restingHeartRate)!,
                HKSampleType.quantityType(forIdentifier: .walkingHeartRateAverage)!,
                HKSampleType.quantityType(forIdentifier: .bodyTemperature)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureSystolic)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureDiastolic)!,
                HKSampleType.quantityType(forIdentifier: .oxygenSaturation)!,
                HKSampleType.quantityType(forIdentifier: .bloodGlucose)!,
                HKSampleType.quantityType(forIdentifier: .electrodermalActivity)!,
                HKSampleType.quantityType(forIdentifier: .bodyMass)!,

                ]))
            healthStore.requestAuthorization(toShare: nil, read: allTypes) { (success, error) in
                if !success {
                    result(false)// Handle the error here.
                } else{
                    result(true)
                }
            }
        } else {
            result(false)// Handle the error here.
        }
        
    } else if(call.method.elementsEqual("getData")){
        let arguments = call.arguments as? NSDictionary
        let index = (arguments?["index"] as? Int) ?? -1
        let startDate = (arguments?["startDate"] as? NSNumber) ?? 0
        let endDate = (arguments?["endDate"] as? NSNumber) ?? 0
        
        
        let dateFrom = Date(timeIntervalSince1970: startDate.doubleValue / 1000)
        let dateTo = Date(timeIntervalSince1970: endDate.doubleValue / 1000)
        
        if #available(iOS 11.0, *) {
            let allTypes = [
                HKSampleType.quantityType(forIdentifier: .bodyFatPercentage)!,
                HKSampleType.quantityType(forIdentifier: .height)!,
                HKSampleType.quantityType(forIdentifier: .bodyMassIndex)!,
                HKSampleType.quantityType(forIdentifier: .waistCircumference)!,
                HKSampleType.quantityType(forIdentifier: .stepCount)!,
                HKSampleType.quantityType(forIdentifier: .basalEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .activeEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .heartRate)!,
                HKSampleType.quantityType(forIdentifier: .bodyTemperature)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureSystolic)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureDiastolic)!,
                HKSampleType.quantityType(forIdentifier: .restingHeartRate)!,
                HKSampleType.quantityType(forIdentifier: .walkingHeartRateAverage)!,
                HKSampleType.quantityType(forIdentifier: .oxygenSaturation)!,
                HKSampleType.quantityType(forIdentifier: .bloodGlucose)!,
                HKSampleType.quantityType(forIdentifier: .electrodermalActivity)!,
                HKSampleType.quantityType(forIdentifier: .bodyMass)!,
                ]
            print("INDEX IS " , index)
            print("COUNT IS " , allTypes.count)
            if(index >= 0 && index < allTypes.count){
                let dataType = allTypes[index]
                print("DATA TYPE IS ", dataType)
                let predicate = HKQuery.predicateForSamples(withStart: dateFrom, end: dateTo, options: .strictStartDate)
                print("PREDICATE ", predicate)
                if (self.healthStore.authorizationStatus(for: dataType) == .sharingAuthorized) {
                    
                }

                let sortDescriptor = NSSortDescriptor(key: HKSampleSortIdentifierEndDate, ascending: true)

                
                let query = HKSampleQuery(sampleType: dataType, predicate: predicate, limit: HKObjectQueryNoLimit, sortDescriptors: [sortDescriptor]) {
                    x, samplesOrNil, error in
                    
                    guard let samples = samplesOrNil as? [HKQuantitySample] else {
                        result(FlutterError(code: "FlutterHealth", message: "Results are null", details: error))
                        return
                    }

                    if(samples != nil){
                    result(samples.map { sample -> NSDictionary in
                        let unit = self.unitFromDartType(type: index)
                        return [
                            "value": sample.quantity.doubleValue(for: unit),
                            "unit": unit.unitString,
                            "date_from": Int(sample.startDate.timeIntervalSince1970 * 1000),
                            "date_to": Int(sample.endDate.timeIntervalSince1970 * 1000),
                            "data_type_index": index,
                        ]
                    })
                    } else {
                        print("Either there are no values or the user did not allow getting this value")
                        result("Either there are no values or the user did not allow getting this value")
                    }
                    return                
                }
                HKHealthStore().execute(query)

            } else{
                print("Something wrong with request")
                result("Unsupported version or data type")
            }
        } else {
            print("Unsupported version 1")
            result("Unsupported version or data type")
        }
        print("Unsupported version")
    }else if(call.method.elementsEqual("getSummary")){
        let arguments = call.arguments as? NSDictionary
        let index = (arguments?["index"] as? Int) ?? -1
        let startDate = (arguments?["startDate"] as? NSNumber) ?? 0
        let endDate = (arguments?["endDate"] as? NSNumber) ?? 0
        
        
        let dateFrom = Date(timeIntervalSince1970: startDate.doubleValue / 1000)
        let dateTo = Date(timeIntervalSince1970: endDate.doubleValue / 1000)

        if #available(iOS 11.0, *) {
            let allTypes = [
                HKSampleType.quantityType(forIdentifier: .bodyFatPercentage)!,
                HKSampleType.quantityType(forIdentifier: .height)!,
                HKSampleType.quantityType(forIdentifier: .bodyMassIndex)!,
                HKSampleType.quantityType(forIdentifier: .waistCircumference)!,
                HKSampleType.quantityType(forIdentifier: .stepCount)!,
                HKSampleType.quantityType(forIdentifier: .basalEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .activeEnergyBurned)!,
                HKSampleType.quantityType(forIdentifier: .heartRate)!,
                HKSampleType.quantityType(forIdentifier: .bodyTemperature)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureSystolic)!,
                HKSampleType.quantityType(forIdentifier: .bloodPressureDiastolic)!,
                HKSampleType.quantityType(forIdentifier: .restingHeartRate)!,
                HKSampleType.quantityType(forIdentifier: .walkingHeartRateAverage)!,
                HKSampleType.quantityType(forIdentifier: .oxygenSaturation)!,
                HKSampleType.quantityType(forIdentifier: .bloodGlucose)!,
                HKSampleType.quantityType(forIdentifier: .electrodermalActivity)!,
                HKSampleType.quantityType(forIdentifier: .bodyMass)!,
                ]
            print("INDEX IN SUMMARY IS " , index)
            print("COUNT IN SUMMARY IS " , allTypes.count)
            if(index >= 0 && index < allTypes.count){
                let dataType = allTypes[index]
                print("DATA TYPE IS ", dataType)

                //  Set the Predicates & Interval
                let predicate = HKQuery.predicateForSamples(withStart: dateFrom, end: dateTo, options: .strictStartDate)
                var interval = DateComponents()
                interval.day = 1

                //  Perform the Query
                let query = HKStatisticsCollectionQuery(quantityType: dataType, quantitySamplePredicate: predicate, options: [.cumulativeSum], anchorDate: dateFrom as Date, intervalComponents:interval)

                query.initialResultsHandler = { query, samplesOrNil, error in

                    guard let samples = samplesOrNil as? [HKQuantitySample] else {
                        result(FlutterError(code: "FlutterHealth", message: "Results are null", details: error))
                        return
                    }
                    print("SAMPLES ARE ", samples)
                    if(samples != nil){
                        result(samples.map { sample -> NSDictionary in
                            let unit = self.unitFromDartType(type: index)
                            print("VALUE IN SUM IS ", sample.quantity.doubleValue(for: unit))
                                return [
                                    "value": sample.quantity.doubleValue(for: unit),
                                    "unit": unit.unitString,
                                    "date_from": Int(sample.startDate.timeIntervalSince1970 * 1000),
                                    "date_to": Int(sample.endDate.timeIntervalSince1970 * 1000),
                                    "data_type_index": index,
                                ]
                        })
                    } else {
                        print("Either there are no values or the user did not allow getting this value")
                        result("Either there are no values or the user did not allow getting this value")
                    }
                    return                
                    HKHealthStore().execute(query)
                }
            } else{
                print("Something wrong with request")
                result("Unsupported version or data type")
            }
        } else {
            print("Unsupported version 1")
            result("Unsupported version or data type")
        }
        print("Unsupported version")                
    }else if(call.method.elementsEqual("getHeartAlerts")){
        let arguments = call.arguments as? NSDictionary
        let index = (arguments?["index"] as? Int) ?? -1
        let startDate = (arguments?["startDate"] as? NSNumber) ?? 0
        let endDate = (arguments?["endDate"] as? NSNumber) ?? 0
        
        
        let dateFrom = Date(timeIntervalSince1970: startDate.doubleValue / 1000)
        let dateTo = Date(timeIntervalSince1970: endDate.doubleValue / 1000)
        
        if #available(iOS 12.2, *) {
            let allTypes = [
                HKSampleType.categoryType(forIdentifier: .highHeartRateEvent)!,
                HKSampleType.categoryType(forIdentifier: .lowHeartRateEvent)!,
                HKSampleType.categoryType(forIdentifier: .irregularHeartRhythmEvent)!,
            ]
            if(index >= 0 && index < allTypes.count){
                let dataType = allTypes[index-16]
                print("DATA TYPE IS ", dataType)
                let predicate = HKQuery.predicateForSamples(withStart: dateFrom, end: dateTo, options: .strictStartDate)
                print("PREDICATE ", predicate)
            
                let sortDescriptor = NSSortDescriptor(key: HKSampleSortIdentifierEndDate, ascending: true)
                
                
                let query = HKSampleQuery(sampleType: dataType, predicate: predicate, limit: HKObjectQueryNoLimit, sortDescriptors: [sortDescriptor]) {
                    x, samplesOrNil, error in
                    
                    guard let samples = samplesOrNil as? [HKQuantitySample] else {
                        result(FlutterError(code: "FlutterHealth", message: "Results are null", details: error))
                        return
                    }
                    
                    if(samples != nil){
                        result(samples.map { sample -> NSDictionary in
                            let unit = self.unitFromDartType(type: index)
                            return [
                                "value": sample.quantity.doubleValue(for: HKUnit.init(from: "count/min")),
                                "unit": unit.unitString,
                                "date_from": Int(sample.startDate.timeIntervalSince1970 * 1000),
                                "date_to": Int(sample.endDate.timeIntervalSince1970 * 1000),
                                "data_type_index": index + 16,
                            ]
                        })
                    } else {
                        print("Either there are no values or the user did not allow getting this value")
                        result("Either there are no values or the user did not allow getting this value")
                    }
                    return
                }
                HKHealthStore().execute(query)
            } else{
                print("Something wrong with request")
                result("Unsupported version or data type")
            }
        } else {
            print("Unsupported version 1")
            result("Unsupported version or data type")
        }
        print("Unsupported version")
        
    }
  }
    

    
    public func unitFromDartType(type: Int) -> HKUnit {
        guard let unit: HKUnit = {
            switch (type) {
            case 0:
                return HKUnit.percent()
            case 1:
                return HKUnit.meter()
            case 2:
                return HKUnit.init(from: "")
            case 3:
                return HKUnit.meter()
            case 4:
                return HKUnit.count()
            case 5,6:
                return HKUnit.kilocalorie()
            case 7, 11, 12:
                return HKUnit.init(from: "count/min")
            case 8:
                return HKUnit.degreeCelsius()
            case 9,10:
                return HKUnit.millimeterOfMercury()
            case 13:
                return HKUnit.percent()
            case 14:
                return HKUnit.init(from: "mg/dl")
            case 15:
                return HKUnit.siemen()
            case 16:
                return HKUnit.gramUnit(with: HKMetricPrefix.kilo)
            default:
                return HKUnit.count()
            }
            }() else {
                return HKUnit.count()
        }
        return unit
    }
}




